src/SharpBrick.PoweredUp.BlueGigaBLE/BlueGigaBLEPoweredUpBluetoothAdapater.cs

```diff
@@ -0,0 +1,514 @@
+        //semaphore for signal successful connection
+        private SemaphoreSlim connected;
+        //semaphore for signal finished discovery of services of a device
+        private SemaphoreSlim allServicesForDeviceDiscovered = new SemaphoreSlim(0, 1);
+        //semaphore for signal finished discovery of characteristics of a service
+        private SemaphoreSlim allCharacteristicsForServiceDiscovered = new SemaphoreSlim(0, 1);
+        private bool wasLastConnectSuccesfull = false;
+        private byte lastConnectionHandle = 0xff;
+        #endregion
+
+        #region Constructor
+        public BlueGigaBLEPoweredUpBluetoothAdapater(String comPortName ,  bool traceDebug, ILogger<SharpBrick.PoweredUp.BlueGigaBLE.BlueGigaBLEPoweredUpBluetoothAdapater> logger = default)
+        {
+            _logger = logger;
+            TraceDebug = traceDebug;
+            Bglib = new Bluegiga.BGLib(_logger , traceDebug);
+            BleStatus = new BlueGigaBLEStatus(BlueGigaBLEStatus.STATE_RESETING, _logger, traceDebug);
+            legoHubServiceBytes = BlueGigaBLEHelper.ConvertUUIDToByteArray(new Guid(PoweredUpBluetoothConstants.LegoHubService));
+            myDevices = new ConcurrentDictionary<ulong, BlueGigaBLEPoweredUpBluetoothDevice>();
+            devicesinfo = new ConcurrentDictionary<ulong, PoweredUpBluetoothDeviceInfo>();
+            //TODO: .NET 5 read the portname from configuration/serviceoptions/parameters /see ServiceCollectionExtensionsForBlueGigaBLE.cs
+            InitSerial(comPortName, 9600);
+            // initialize BGLib events needed for the Adapter
+            //fires on reset/reboot
+            Bglib.BLEEventSystemBoot += new Bluegiga.BLE.Events.System.BootEventHandler(this.SystemBootEventHandler);
+            //fires on Connect-trial
+            Bglib.BLEEventConnectionStatus += new Bluegiga.BLE.Events.Connection.StatusEventHandler(ConnectionStatusHandler);
+            //fires on finding a service
+            Bglib.BLEEventATTClientGroupFound += new Bluegiga.BLE.Events.ATTClient.GroupFoundEventHandler(ServiceFoundHandler);
+            
+            //fires on finding an attribute/characteristic
+            Bglib.BLEEventATTClientFindInformationFound += new Bluegiga.BLE.Events.ATTClient.FindInformationFoundEventHandler(CharacteristicFoundHandler);
+            //fires on end of service-discovery or end of characteristic-discovery
+            Bglib.BLEEventATTClientProcedureCompleted += new Bluegiga.BLE.Events.ATTClient.ProcedureCompletedEventHandler(ServiceOrCharacteristicSearchCompletedHandler);
+            Byte[] cmd = Bglib.BLECommandSystemReset(0x00); //boot normal, not in Device-Firmware-Update-mode!
+        }
+        #endregion
+        #region Initialization and System-/Serial-handler
+        private void InitSerial(String portname, int baudrate)
+        {
+            this.SerialAPI = new System.IO.Ports.SerialPort();
+            SerialAPI.Handshake = System.IO.Ports.Handshake.RequestToSend;
```

Comment:
Property Initalizers ;) still guessing your background ;)

Reply:
often results of trial and error lead me to such things :-)
changed that; see in next push to the PR

Reply:
I took out the InitSerial and put it back into the constructor and simplified the code...